src/libopenrave-core/jsonparser/jsonreader.cpp

```diff
@@ -298,6 +298,36 @@ class JSONReader
                 }
                 return false;
             }
+            // if parent rEnvInfo is going to overwrite bodies in the environment of the referenceUri, do not deserialize them. otherwise, they would be merged instead of overwritten.
+            if( rEnvInfo.HasMember("bodies") && rEnvInfo["bodies"].Size() > 0 ) {
+                if( prReferenceEnvInfo->HasMember("bodies") && (*prReferenceEnvInfo)["bodies"].Size() > 0 ) {
+                    boost::shared_ptr<rapidjson::Document> prFilteredReferenceEnvInfo = boost::make_shared<rapidjson::Document>(&alloc);
+                    prFilteredReferenceEnvInfo->CopyFrom(*prReferenceEnvInfo, alloc);
+                    const rapidjson::Value& rEnvBodies = rEnvInfo["bodies"];
+                    rapidjson::Value& rRefBodies = (*prFilteredReferenceEnvInfo)["bodies"];
+                    for (rapidjson::Value::ConstValueIterator itEnvBody = rEnvBodies.Begin(); itEnvBody != rEnvBodies.End(); ++itEnvBody) {
+                        rapidjson::Value::ConstMemberIterator itEnvBodyId = itEnvBody->FindMember("id");
```

Comment:
Are we also matching name here? Or only `id`?

Should we be assigning a new `id` in an attempt to keep both bodies?

Reply:
If someone had a dynamic body with id=1, and later go into webstack and change static scene to have a body with id=1 as well, and reload the process, should both bodies be present?

Reply:
@ziyan If we think about modifications on the referenced environment, there are a few conditions to consider. See below table. I am not certain about expected behaviors on condition 2-5. Can I get advice? @ziyan @rdiankov 


| no | condition                                                            | action                                                                                                                           |
|----|----------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------|
| 1  | name and id match                                                    | no doubt! replace the body                                                                                                       |
| 2  | name matches. but id doesn't. there is no body that has the same id  | maybe the body in reference env was replaced. replace the body?                                                                  |
| 3  | name matches. but id doesn't. there is a body that has the same id   | maybe the body in reference env was named and a new body was added with the original name. raise an error or discard the change in dynamic state? |
| 4  | id matches but name doesn't. there is no body that has the same name | maybe the body was renamed. replace the body with the name in reference env?                                                     |
| 5  | id matches but name doesn't. there is a body that has the same name  | same as no. 3                                                                                                                       |
| 6  | nothing matches                                                      | create a new body.                                                                                                               |

Reply:
updated the PR according to the above table for now.

Reply:
in 1, "replace the body", is this the right thing to do? @rschlaikjer what's the behavior right now if we use `NotifyEnvironmentChanged` to modify a body that exists in the static scene to give it extra readable interface and persist this in the dynamic scene? 

in my opinion, if `id` exists, it should always take priority, `name` is just a property that can be modified by the JSON to be applied in this case. Only if `id` does not exist, or is empty, OpenRAVE has this fallback behavior of using `name` in its place. This behavior does not exist in webstack. Doing all 6 cases is super confusing.

Reply:
@ziyan `NotifyEnvironmentChanged`  does not use different code path and it applies update upon an existing body in the current environment as it does not have 'referenceUri`.

The above table is taking `id` priority.. In addition, in case `id` does not match, it falls back to use `name` if `possible (no 2). and if prioritizing `id` leads to generate duplicated `name`s, raises an error. (no 3 and 5).

Reply:
Id matches, name matches: readable is added
https://pastebin.mujin.co.jp/?f72214fe53d642fb#6a2pGqRvRWh4sBt3cQHB39vkLwrnkarMQ58SGuhsXC1e

Id matches, name different: body is renamed, readable is added
https://pastebin.mujin.co.jp/?52c64fd37cd60439#FdXpwr8DkBUsDPH7Bhduhr8na9pQppb6h39HNjQBFvva

Id doesn't match, name matches: id is changed (??), readable is added
https://pastebin.mujin.co.jp/?d1c0b49fd4988e9a#2z4MsGqD1K8AFFod3ZUwLGHYgU342guX22cJqCrsaYCW

Id doesn't match, name doesn't match: adds new body
https://pastebin.mujin.co.jp/?091d610e9d8fa804#9s8LMJekJDNFvs2VehGv7G72aknboaNbCMF1HCmeaTbD

Reply:
@ziyan Any more comments?

Reply:
@yoshikikanemoto No more comments, thank you!